### PR TITLE
Fix issue with catching exceptions thrown by engage

### LINKED ISSUE #44: `handleWith` causes Future chain to be called twice if `engage` `resolve` fails.
This is pretty edge case and strange behavior. This was discovered while using the library where the Future in question changed test state. When an assertion in the `.engage` on that future failed, it caused the `handleWith` code of the already engaged Future to be called again, which cause the map to be called again, which then blew up because the state had already changed. I pushed a minimal reproduction of the bug on the branch `handle-called-out-of-scope`.

```typescript
test("does not get run if engages above this scope throw", (done) => {
            let mapCalledTimes = 0;
            let handleWithCalledTimes = 0;
            const action = Future.of(33)
                .handleWith((e) => {
                    console.log(`failed an infallible future: ${e}`);
                    handleWithCalledTimes++;
                    return Future.of(-1);
                })
                .map((r) => {
                    mapCalledTimes++;
                    return r;
                });

            try {
                action.engage(
                    (e) => {
                        throw e;
                    },
                    (r) => {
                        throw new Error(`oh no, something went wrong after the future has run to completion on ${r}`);
                    }
                );
            } catch (e) {
                expect(handleWithCalledTimes).toBe(0);
                expect(mapCalledTimes).toBe(1);
                done();
            }
        });
```

That's the test entirely though. The error thrown in the `.engage` causes the whole chain of the Future the `engage` was called on to run when it's somehow caught by the `handleWith`. The same thing happens if you start with a rejected Future, though then everything in the chain is run twice including the `handleWith`.

### PR DESCRIPTION
**Pull request by coltfred**
This fixes #44. The test that's commented out is one that I don't believe should pass, but I'm open to discussion on it.

I would really like a more elegant solution, but I didn't find one that didn't result in an entire rewrite of the class besides this. I'm open to pursuing different ideas if we think this doesn't fix it appropriately. Also totally open to additional test cases.

### CODE REVIEW

**Comment by leeroy-travis** [comment]
Commit SHA:419801d2031de141302a115ee49c7df3286512ec
No changes to code coverage between the base branch and the head branch

**Comment by leeroy-travis** [comment]
Commit SHA:26620d6af0c423cc2ca8015915c97b1cd61e9942
No changes to code coverage between the base branch and the head branch

File: Future.test.ts [hunk 1]
```diff
@@ -37,6 +37,69 @@ describe("Future", () => {
 
             expect(resolveSpy).toHaveBeenCalledWith("my value");
         });
+        test("does not get run if engages above this scope throw", (done) => {
```
Comment (skeet70) [inline Future.test.ts:40]:
```suggestion
        test("handleWith for current Future does not get run if errors above its scope throw", (done) => {
```

File: Future.ts [hunk 2]
```diff
@@ -43,7 +50,7 @@ export default class Future<L, R> {
      * @return {Promise<R>} Start execution of the Future but return a Promise which will be resolved/reject when the Future is
      */
     toPromise(): Promise<R> {
-        return new Promise<R>((resolve: Resolve<R>, reject: Reject<L>) => this.engage(reject, resolve));
+        return new Promise<R>((resolve: Resolve<R>, reject: Reject<L>) => this.engageAndCatch(reject, resolve));
```
Comment (skeet70) [inline Future.ts:53]: I think this can be `engage`? A throw from inside a Promise like this should get caught by the promise. Actually why does `engageAndCatch` exist with the `safe` flag like this at all? If we want it, can't we have `noThrowEngage` or `engageAndCatch` without the `safe` argument and always do the `safe` logic? Anywhere you would write `engageAndCatch(x, y, false)` you can write `engage(x, y)` instead right?

**Review by skeet70 — CHANGES_REQUESTED** [review]

File: Future.ts (hunk 2 above)
Comment (coltfred) [reply Future.ts:53]: You're right on both accounts. Fixed and added a test that shows that an error thrown in the future will be caught by the promise if in a chain.

**Review by skeet70 — APPROVED** [review]
